### PR TITLE
[RW-6272][risk=low][Spring 2 Upgrade Step 5] The rest of Spring upgrade task

### PR DESCRIPTION
**Pull request by yonghaoy**
Description:

- Most of them are JPA updates
- Some dependencies update
- application.properities config updates
- Integration test and BigQuery Test fixes 
Some more stuff: 
https://docs.google.com/document/d/1dGkLj6duvC39lfrasbChU7Ujv1HNGwYbtg2QYWSOj_w/edit?resourcekey=0-H7kyKq7s_CiQpW6oaJJCNg#

---
**PR checklist**

- [ ] This PR meets the Acceptance Criteria in the JIRA story
- [ ] The JIRA story has been moved to Dev Review
- [ ] This PR includes appropriate unit tests
- [ ] I have run and tested this change locally
- [ ] I have run the E2E tests on ths change against my local UI and/or API server with `yarn test-local` or [yarn test-local-devup](https://github.com/all-of-us/workbench/blob/master/e2e/README.md#examples)
- [ ] If this includes a UI change, I have taken screen recordings or screenshots of the new behavior and notified the PO and UX designer
- [ ] If this includes an API change, I have updated the appropriate Swagger definitions and notified API consumers
- [ ] If this includes a new feature flag, I have created and linked new JIRA tickets to (a) turn on the feature flag and (b) remove it later

### CODE REVIEW

File: api/src/main/java/org/pmiops/workbench/firecloud/FireCloudConfig.java [hunk 1]
```diff
@@ -220,10 +219,4 @@ public static ApiClient buildApiClient(WorkbenchConfig workbenchConfig) {
         .setReadTimeout(workbenchConfig.firecloud.timeoutInSeconds, TimeUnit.SECONDS);
     return apiClient;
   }
-
-  @Bean
-  @Lazy
-  public IamCredentialsClient getIamCredentialsClient() throws IOException {
```
Comment (yonghaoy) [inline api/src/main/java/org/pmiops/workbench/firecloud/FireCloudConfig.java:226]: It is a duplicated bean that Spring2 does not like. 
Feels it make sense to use another one:
https://github.com/all-of-us/workbench/blob/master/api/src/main/java/org/pmiops/workbench/google/GoogleConfig.java

File: api/tools/src/main/java/org/pmiops/workbench/tools/CommandLineToolApplicationRunListener.java [hunk 2]
```diff
@@ -48,7 +48,4 @@ public void contextPrepared(ConfigurableApplicationContext context) {
 
   @Override
   public void contextLoaded(ConfigurableApplicationContext context) {}
-
-  @Override
-  public void finished(ConfigurableApplicationContext context, Throwable exception) {}
```
Comment (yonghaoy) [inline api/tools/src/main/java/org/pmiops/workbench/tools/CommandLineToolApplicationRunListener.java:53]: no such method anymore

File: api/src/main/resources/application.properties [hunk 3]
```diff
@@ -13,16 +13,21 @@
...
 spring.jpa.hibernate.ddl-auto=none
+spring.liquibase.enabled=false
 spring.jpa.properties.hibernate.dialect=org.pmiops.workbench.cdr.MySQLDialect
-# See https://docs.spring.io/spring-boot/docs/1.5.3.RELEASE/reference/html/production-ready-endpoints.html
+# See https://docs.spring.io/spring-boot/docs/2.3.9.RELEASE/reference/html/production-ready-features.html
 # and https://precisionmedicineinitiative.atlassian.net/browse/RW-1478
 # Disable all spring-boot endpoints unless specifically necessary
-endpoints.enabled=false
+management.endpoints.enabled-by-default=false
+# Restore 1.x Actuator feature.
```
Comment (calbach) [inline api/src/main/resources/application.properties:31]: Is there a way to just disable this?

File: api/src/test/resources/application.properties [hunk 4]
```diff
@@ -4,16 +4,16 @@ spring.datasource.username=sa
 spring.datasource.password=sa
 spring.jpa.hibernate.ddl-auto=create-drop
 spring.jpa.properties.hibernate.dialect=org.pmiops.workbench.cdr.CommonTestDialect
-liquibase.enabled=false
+spring.liquibase.enabled=false
 # To allow each tests override and customize their the beans.
 spring.main.allow-bean-definition-overriding=true
 # Uncomment the following to turn on full SQL debugging
-#spring.jpa.properties.hibernate.show_sql=true
+spring.jpa.properties.hibernate.show_sql=true
```
Comment (calbach) [inline api/src/test/resources/application.properties:11]: nit: revert comment lines?

File: api/build.gradle [hunk 5]
```diff
@@ -428,34 +433,38 @@ dependencies {
...
-  compile("org.springframework.boot:spring-boot-starter-web:$project.ext.SPRING_BOOT_VERSION") {
+  compile("org.springframework.boot:spring-boot-starter-web") {
     exclude module: 'spring-boot-starter-tomcat'
     exclude group: 'org.slf4j', module: 'jul-to-slf4j'
   }
 
-  compile "org.springframework.retry:spring-retry:1.1.5.RELEASE"
-  compile "org.springframework.security:spring-security-core:$project.ext.SPRING_SECURITY_VERSION"
-  compile "org.springframework.security:spring-security-web:$project.ext.SPRING_SECURITY_VERSION"
+  compile "org.springframework.retry:spring-retry:1.3.1"
```
Comment (calbach) [inline api/build.gradle:454]: Does this one not automatically inherit the version like the other spring packages?

File: api/build.gradle [hunk 6]
```diff
@@ -428,34 +433,38 @@ dependencies {
 
   compile "org.json:json:20160810"
 
-  compile("org.springframework.boot:spring-boot-starter-actuator:$project.ext.SPRING_BOOT_VERSION") {
+  compile("org.apache.tomcat:tomcat-jdbc:9.0.10")
+  compile("org.springframework.boot:spring-boot-starter-tomcat")
+  compile("org.springframework.boot:spring-boot-starter-jdbc")
+  compile("org.springframework.boot:spring-boot-starter-actuator") {
```
Comment (calbach) [inline api/build.gradle:439]: Do we need this one?

File: api/src/test/java/org/pmiops/workbench/tools/cdrconfig/CdrConfigVOMapperTest.java [hunk 7]
```diff
@@ -23,7 +24,7 @@
 import org.springframework.test.context.junit4.SpringRunner;
 
 @RunWith(SpringRunner.class)
-@Import(CdrConfigVOMapperImpl.class)
+@Import({CdrConfigVOMapperImpl.class, CommonConfig.class})
 @DataJpaTest
 @DirtiesContext(classMode = DirtiesContext.ClassMode.BEFORE_EACH_TEST_METHOD)
 public class CdrConfigVOMapperTest extends SpringTest {
```
Comment (calbach) [inline api/src/test/java/org/pmiops/workbench/tools/cdrconfig/CdrConfigVOMapperTest.java:30]: Withe the above change, it seems like there is no need to extend `SpringTest` anymore - can that be deleted?

**Review by calbach — COMMENTED** [review]
Nice - thanks for working through all these issues. I will pull this down  locally and poke around a bit.

File: api/src/main/java/org/pmiops/workbench/cohorts/CohortService.java [hunk 8]
```diff
@@ -31,7 +31,7 @@ public CohortService(
   }
 
   public List<Cohort> findAll(List<Long> cohortIds) {
-    return ((List<DbCohort>) cohortDao.findAll(cohortIds))
+    return ((List<DbCohort>) cohortDao.findAllById(cohortIds))
```
Comment (jmthibault79) [inline api/src/main/java/org/pmiops/workbench/cohorts/CohortService.java:34]: this is a sketchy cast - we should remove it as well as the cast in ConceptService

File: api/src/main/java/org/pmiops/workbench/db/dao/UserServiceImpl.java [hunk 9]
```diff
@@ -155,7 +155,7 @@ public DbUser updateUserWithRetries(
         return userDao.save(dbUser);
       } catch (ObjectOptimisticLockingFailureException e) {
         if (objectLockingFailureCount < MAX_RETRIES) {
-          dbUser = userDao.findOne(dbUser.getUserId());
+          dbUser = userDao.findById(dbUser.getUserId()).orElse(null);
```
Comment (jmthibault79) [inline api/src/main/java/org/pmiops/workbench/db/dao/UserServiceImpl.java:158]: null would be bad here so these should use orElseThrow()

File: api/src/main/resources/application.properties (hunk 3 above)
Comment (yonghaoy) [reply api/src/main/resources/application.properties:31]: The Actuator? I am not sure what they are, and if we are using those endpoints actually. I was trying to keep 1 and 2 endpoints path the same:
https://docs.spring.io/spring-boot/docs/current/reference/html/production-ready-features.html
If we don't need this, I think I can remove it.

File: api/build.gradle (hunk 5 above)
Comment (yonghaoy) [reply api/build.gradle]: oh, right, I installed spring dependency manager now. Giving a shot, that one's version is slightly different than others as the latest version is just 1.3

File: api/build.gradle (hunk 6 above)
Comment (yonghaoy) [reply api/build.gradle:439]: That one was auto imported by some tool I remembeted. Let me try to remove this..

File: api/src/main/java/org/pmiops/workbench/cohorts/CohortService.java (hunk 8 above)
Comment (yonghaoy) [reply api/src/main/java/org/pmiops/workbench/cohorts/CohortService.java:34]: right, that does seems not good. Changed

**Review by calbach — APPROVED** [review]
I pulled locally and everything appears to be working well - nice! Aside from my comment, the only issue I saw was the following when starting up:

Initially, I got a almost never-ending stream of updates from the hot-swapping task. Finally this converged after seemingly updating every class in the system - about 5m. This might be avoidable with a one-time command (maybe clean or something) when doing the upgrade, but something to keep an eye on..

**Comment by yonghaoy** [comment]
> Nice - thanks for working through all these issues. I will pull this down locally and poke around a bit.

Tried some functions in local server, seems working as usual

File: api/src/main/resources/application.properties [hunk 10]
```diff
@@ -13,16 +13,21 @@
...
 # see in Spring documentation) will do nothing. See WorkbenchDbConfig and
 # CdrDbConfig.
-spring.datasource.test-on-borrow=true
-spring.datasource.test-while-idle=true
-spring.datasource.time-between-eviction-runs-millis=3600000
-spring.datasource.validation-query=SELECT 1
-spring.datasource.initial-size=5
-spring.datasource.max-idle=5
-spring.datasource.min-idle=1
+spring.datasource.type=org.apache.tomcat.jdbc.pool.DataSource
```
[the remark this answers is not in the document]
Comment (yonghaoy) [reply api/src/main/resources/application.properties:16]: Done, also exclude it from `spring-boot-starter-jdbc`

**Review by freemabd — COMMENTED** [review]
Initial pass everything looks fine. I feel that we should wait to merge this PR after the release on Monday. This gives us a week in test env to work out any kinks before releasing this to prod.

**Comment by freemabd** [comment]
Thanks for all the hard work in this @yonghaoy. Much appreciated!

File: api/src/main/java/org/pmiops/workbench/db/dao/UserServiceImpl.java [hunk 11]
```diff
@@ -170,7 +177,7 @@ public DbUser updateUserWithRetries(
             .getMessage()
             .equals("Statement closed.")) {
           if (statementClosedCount < MAX_RETRIES) {
-            dbUser = userDao.findOne(dbUser.getUserId());
+            dbUser = userDao.findById(dbUser.getUserId()).orElse(null);
```
Comment (jmthibault79) [inline api/src/main/java/org/pmiops/workbench/db/dao/UserServiceImpl.java:180]: this should also be orElseThrow

**Comment by yonghaoy** [comment]
> Initial pass everything looks fine. I feel that we should wait to merge this PR after the release on Monday. This gives us a week in test env to work out any kinks before releasing this to prod.

Thanks! That is a good idea. I will hold until after Monday's release to have this PR

**Pull request merged**